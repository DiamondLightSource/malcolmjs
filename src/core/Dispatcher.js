/*
 * React.js Starter Kit
 * Copyright (c) 2014 Konstantin Tarkus (@koistya), KriaSoft LLC.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE.txt file in the root directory of this source tree.
 */

import Flux from 'flux';
import PayloadSources from '../constants/PayloadSources';
import assign from 'react/lib/Object.assign';

/**
 * A singleton that operates as the central hub for application updates.
 * For more information visit https://facebook.github.io/flux/
 */
let Dispatcher = assign(new Flux.Dispatcher(), {

  /**
   * @param {object} action The details of the action, including the action's
   * type and additional data coming from the server.
   */
  handleServerAction(action) {
    var payload = {
      source: PayloadSources.SERVER_ACTION,
      action: action
    };
    this.dispatch(payload);
  },

  /**
   * @param {object} action The details of the action, including the action's
   * type and additional data coming from the view.
   */
  handleViewAction(action) {
    var payload = {
      source: PayloadSources.VIEW_ACTION,
      action: action
    };
    this.dispatch(payload);
  }

});

export default Dispatcher;
