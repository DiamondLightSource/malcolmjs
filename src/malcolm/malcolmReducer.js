import {
  MalcolmNewBlock,
  MalcolmSend,
  MalcolmError,
  MalcolmBlockMeta,
  MalcolmAttributeData,
} from './malcolm.types';

const initialMalcolmState = {
  messagesInFlight: [],
  blocks: {},
  parentBlock: undefined,
};

function updateMessagesInFlight(state, action) {
  const newState = state;
  newState.messagesInFlight = [
    ...state.messagesInFlight,
    { ...action.payload },
  ];
  return newState;
}

function stopTrackingMessage(state, action) {
  return {
    ...state,
    messagesInFlight: state.messagesInFlight.filter(
      m => m.id !== action.payload.id
    ),
  };
}

function registerNewBlock(state, action) {
  const blocks = { ...state.blocks };
  blocks[action.payload.blockName] = {
    name: action.payload.blockName,
    loading: true,
  };

  return {
    ...state,
    blocks,
    parentBlock: action.payload.parent
      ? action.payload.blockName
      : state.parentBlock,
  };
}

function updateBlock(state, payload) {
  const blocks = { ...state.blocks };

  if (payload.delta) {
    const blockName = state.messagesInFlight.find(m => m.id === payload.id)
      .path[0];

    if (Object.prototype.hasOwnProperty.call(blocks, blockName)) {
      blocks[blockName] = {
        ...blocks[blockName],
        loading: false,
        label: payload.label,
        attributes: payload.fields.map(f => ({ name: f, loading: true })),
      };
    }
  }

  return {
    ...state,
    blocks,
  };
}

function updateAttribute(state, payload) {
  if (payload.delta) {
    const { path } = state.messagesInFlight.find(m => m.id === payload.id);
    const blockName = path[0];
    const attributeName = path[1];

    if (Object.prototype.hasOwnProperty.call(state.blocks, blockName)) {
      const attributes = [...state.blocks[blockName].attributes];

      const matchingAttribute = attributes.findIndex(
        a => a.name === attributeName
      );
      if (matchingAttribute >= 0) {
        attributes[matchingAttribute] = {
          ...attributes[matchingAttribute],
          loading: false,
        };
      }

      const blocks = { ...state.blocks };
      blocks[blockName] = { ...state.blocks[blockName], attributes };

      return {
        ...state,
        blocks,
      };
    }
  }

  return state;
}

const malcolmReducer = (state = initialMalcolmState, action) => {
  switch (action.type) {
    case MalcolmNewBlock:
      return registerNewBlock(state, action);

    case MalcolmSend:
      return updateMessagesInFlight(state, action);

    case MalcolmError:
      return stopTrackingMessage(state, action);

    case MalcolmBlockMeta:
      return updateBlock(state, action.payload);

    case MalcolmAttributeData:
      return updateAttribute(state, action.payload);

    default:
      return state;
  }
};

export default malcolmReducer;
