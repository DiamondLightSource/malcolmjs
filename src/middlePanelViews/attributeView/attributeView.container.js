import React from 'react';
import { connect } from 'react-redux';
import PropTypes from 'prop-types';
import { withTheme } from '@material-ui/core/styles';
import Typography from '@material-ui/core/Typography';
import Loadable from 'react-loadable';
import TabbedPanel from '../tabbedMiddlePanel.component';
import ArchiveTable from './archiveTable.container';
import {
  AlarmStates,
  AlarmStatesByIndex,
} from '../../malcolmWidgets/attributeDetails/attributeAlarm/attributeAlarm.component';

const LoadablePlotter = Loadable({
  loader: () => import('../plotter.component'),
  loading: () => <Typography>Loading...</Typography>,
});

const updatePlotData = (oldDataElement, alarmIndex, attribute) => {
  const dataElement = oldDataElement;
  const alarms = attribute.alarmState.toarray();
  dataElement.x = attribute.timeStamp.toarray();
  dataElement.y = attribute.plotValue
    .toarray()
    .map(
      (value, valIndex) =>
        alarms[valIndex] === AlarmStatesByIndex[alarmIndex] ||
        (alarms[valIndex - 1] === AlarmStatesByIndex[alarmIndex] &&
          alarms[valIndex] !== AlarmStates.UNDEFINED_ALARM)
          ? value
          : null
    );
  dataElement.visible = dataElement.y.some(val => val !== null);

  return dataElement;
};

export const deriveStateFromProps = (props, state) => {
  const { data, layout } = state;
  const dataRevision = `${props.attribute.parent}#${props.attribute.name}#${
    props.attribute.plotTime
  }#p:${props.parentPanelOpen}#c:${props.childPanelOpen}`;
  if (props.attribute && layout.datarevision !== dataRevision) {
    const newData = data.map((dataElement, index) =>
      updatePlotData(dataElement, index, props.attribute)
    );
    layout.datarevision = dataRevision;
    if (
      props.attribute.parent !== layout.parent ||
      props.attribute.name !== layout.attribute
    ) {
      layout.parent = props.attribute.parent;
      layout.attribute = props.attribute.name;
      layout.yaxis = {
        color: props.theme.palette.text.primary,
      };
    }
    return {
      ...state,
      data: [...newData],
      layout,
    };
  }
  return state;
};

const TAB_LABELS = ['Table', 'Plot'];

const AttributeViewer = props => (
  <TabbedPanel tabLabels={TAB_LABELS}>
    <ArchiveTable
      blockName={props.blockName}
      attributeName={props.attributeName}
    />
    <LoadablePlotter
      blockName={props.blockName}
      attributeName={props.attributeName}
      deriveState={deriveStateFromProps}
      doTick
    />
  </TabbedPanel>
);

const mapStateToProps = () => ({});

AttributeViewer.propTypes = {
  blockName: PropTypes.string.isRequired,
  attributeName: PropTypes.string.isRequired,
};

export default connect(mapStateToProps)(withTheme()(AttributeViewer));
