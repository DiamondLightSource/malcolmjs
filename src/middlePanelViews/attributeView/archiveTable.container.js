import React from 'react';
import { connect } from 'react-redux';
import PropTypes from 'prop-types';
import WidgetTable from '../../malcolmWidgets/table/virtualizedTable.component';
import { ARCHIVE_REFRESH_INTERVAL } from '../../malcolm/reducer/malcolmReducer';
import blockUtils from '../../malcolm/blockUtils';

const noOp = () => {};

class ArchiveTable extends React.Component {
  static getDerivedStateFromProps(props, state) {
    if (
      props.attribute.plotTime - state.renderTime >
      ARCHIVE_REFRESH_INTERVAL
    ) {
      return { renderTime: props.attribute.plotTime };
    }
    return state;
  }

  constructor(props) {
    super(props);
    this.state = { renderTime: -1 };
  }

  shouldComponentUpdate(nextProps) {
    return (
      nextProps.attribute.plotTime - this.state.renderTime >
        ARCHIVE_REFRESH_INTERVAL ||
      nextProps.attribute.parent !== this.props.attribute.parent ||
      nextProps.attribute.name !== this.props.attribute.name
    );
  }
  render() {
    const { attribute } = this.props;
    return (
      <WidgetTable
        attribute={{
          raw: {
            value: {
              timeStamp: attribute.timeStamp
                .toarray()
                .slice(0, -1)
                .map(date => date.toISOString()),
              value: attribute.value.toarray(), // .slice(-100),
              alarm: attribute.alarmState.toarray().slice(0, -1),
            },
            meta: {
              elements: {
                alarm: {
                  tags: ['info:alarm'],
                  label: ' ',
                },
                timeStamp: {
                  tags: ['widget:textupdate'],
                  label: 'Time set',
                },
                value: {
                  tags: ['widget:textupdate'],
                  label: 'Value',
                },
              },
            },
          },
          calculated: {},
        }}
        hideInfo
        eventHandler={noOp}
        setFlag={noOp}
        addRow={noOp}
        infoClickHandler={noOp}
        rowClickHandler={noOp}
      />
    );
  }
}

const mapStateToProps = (state, ownProps) => {
  let attribute;
  if (ownProps.attributeName && ownProps.blockName) {
    const attributeIndex = blockUtils.findAttributeIndex(
      state.malcolm.blocks,
      ownProps.blockName,
      ownProps.attributeName
    );
    if (attributeIndex !== -1) {
      attribute =
        state.malcolm.blockArchive[ownProps.blockName].attributes[
          attributeIndex
        ];
    }
  }
  return {
    attribute,
  };
};

ArchiveTable.propTypes = {
  attribute: PropTypes.shape({
    parent: PropTypes.string,
    name: PropTypes.string,
    value: PropTypes.arrayOf(PropTypes.number),
    timeSinceConnect: PropTypes.arrayOf(PropTypes.number),
    plotTime: PropTypes.number,
    isBool: PropTypes.bool,
  }).isRequired,
};

export default connect(mapStateToProps)(ArchiveTable);
