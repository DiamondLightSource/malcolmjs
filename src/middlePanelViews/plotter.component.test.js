import React from 'react';
import { createShallow, createMount } from '@material-ui/core/test-utils';
import MockCircularBuffer from '../malcolm/reducer/attribute.reducer.mocks';
import {
  deriveStateFromProps as deriveMethodState,
  dummyArchive,
} from './methodView/methodArchive.container';
import { deriveStateFromProps as deriveAttributeState } from './attributeView/attributeView.container';
import { malcolmTypes } from '../malcolmWidgets/attributeDetails/attributeSelector/attributeSelector.component';
import { MalcolmTickArchive } from '../malcolm/malcolm.types';
import Plotter, {
  comparePlotlyDateString,
  plotlyDateFormatter,
} from './plotter.component';
import { ARCHIVE_REFRESH_INTERVAL } from '../malcolm/reducer/malcolmReducer';

jest.useFakeTimers();

const mockTheme = {
  palette: {
    primary: { light: '#f0f0f0' },
    text: { primary: '#000' },
    background: { default: '#424242' },
  },
  alarmState: { warning: '#fff', error: '#ff0000', disconnected: '#0500ff' },
};

describe('MethodPlot', () => {
  let shallow;
  let testArchive;

  beforeEach(() => {
    shallow = createShallow({ dive: true });
    testArchive = {
      value: new MockCircularBuffer(5),
      timeStamp: new MockCircularBuffer(5),
      alarmState: new MockCircularBuffer(5),
    };
    testArchive.value.push({
      runParameters: { test1: 0 },
      returned: { test2: 1 },
    });
    testArchive.timeStamp.push({
      localRunTime: new Date(0),
      localReturnTime: new Date(4),
    });
    testArchive.alarmState.push(0);
    testArchive.value.push({
      runParameters: { test1: 0 },
      returned: { test2: 1 },
    });
    testArchive.timeStamp.push({
      localRunTime: new Date(2),
      localReturnTime: new Date(5),
    });
    testArchive.alarmState.push(1);
    testArchive.value.push({
      runParameters: { test1: 0 },
      returned: { test2: 1 },
    });
    testArchive.timeStamp.push({
      localRunTime: new Date(3),
      localReturnTime: new Date(6),
    });
    testArchive.alarmState.push(2);
  });

  it('renders correctly for takes', () => {
    const wrapper = shallow(
      <Plotter
        store={{
          getState: () => {},
          subscribe: () => {},
        }}
        attribute={dummyArchive(testArchive, ['takes', 'test1'])}
        openPanels={{}}
        theme={mockTheme}
        deriveState={deriveMethodState}
      />
    );
    expect(wrapper.dive()).toMatchSnapshot();
  });

  it('renders correctly for returns', () => {
    const wrapper = shallow(
      <Plotter
        store={{
          getState: () => {},
          subscribe: () => {},
        }}
        attribute={dummyArchive(testArchive, ['returns', 'test2'])}
        openPanels={{}}
        theme={mockTheme}
        deriveState={deriveMethodState}
      />
    );
    expect(wrapper.dive()).toMatchSnapshot();
  });
});

describe('attributePlot', () => {
  let shallow;
  let mount;
  let actions;
  let mockArchive;
  let mockStore;

  beforeEach(() => {
    jest.runAllTimers();
    actions = [];
    shallow = createShallow({ dive: true });
    mount = createMount();
    mockArchive = {
      parent: 'test1',
      name: 'attr1',
      timeStamp: new MockCircularBuffer(5),
      alarmState: new MockCircularBuffer(5),
      plotValue: new MockCircularBuffer(5),
      meta: {},
    };
    mockStore = {
      getState: () => {},
      subscribe: () => {},
      dispatch: action => {
        actions.push(action);
      },
    };
  });

  it('renders correctly for number', () => {
    mockArchive.meta.typeid = malcolmTypes.number;
    mockArchive.timeStamp.push(new Date(1000));
    mockArchive.timeStamp.push(new Date(2000));
    mockArchive.timeStamp.push(new Date(3000));
    mockArchive.plotValue.push(1);
    mockArchive.plotValue.push(-1);
    mockArchive.plotValue.push(2);
    mockArchive.alarmState.push(0);
    mockArchive.alarmState.push(0);
    mockArchive.alarmState.push(0);

    const wrapper = shallow(
      <Plotter
        store={mockStore}
        attribute={mockArchive}
        openPanels={{}}
        theme={mockTheme}
        deriveState={deriveAttributeState}
        doTick
      />
    );
    expect(wrapper.dive()).toMatchSnapshot();
  });

  it('renders correctly for bool', () => {
    mockArchive.meta.typeid = malcolmTypes.bool;
    mockArchive.timeStamp.push(new Date(1000));
    mockArchive.timeStamp.push(new Date(2000));
    mockArchive.timeStamp.push(new Date(3000));
    mockArchive.plotValue.push(1);
    mockArchive.plotValue.push(0);
    mockArchive.plotValue.push(1);
    mockArchive.alarmState.push(0);
    mockArchive.alarmState.push(0);
    mockArchive.alarmState.push(0);

    const wrapper = shallow(
      <Plotter
        store={mockStore}
        attribute={mockArchive}
        openPanels={{}}
        theme={mockTheme}
        deriveState={deriveAttributeState}
        doTick
      />
    );
    expect(wrapper.dive()).toMatchSnapshot();
  });

  it('renders correctly for changing alarm state', () => {
    mockArchive.meta.typeid = malcolmTypes.number;
    mockArchive.timeStamp.push(new Date(1000));
    mockArchive.timeStamp.push(new Date(2000));
    mockArchive.timeStamp.push(new Date(3000));
    mockArchive.plotValue.push(1);
    mockArchive.plotValue.push(-1);
    mockArchive.plotValue.push(2);
    mockArchive.alarmState.push(0);
    mockArchive.alarmState.push(1);
    mockArchive.alarmState.push(0);

    const wrapper = shallow(
      <Plotter
        store={mockStore}
        attribute={mockArchive}
        openPanels={{}}
        theme={mockTheme}
        deriveState={deriveAttributeState}
        doTick
      />
    );
    expect(wrapper.dive()).toMatchSnapshot();
  });

  it('dispatches tick action after timeout', () => {
    mockArchive.meta.typeid = malcolmTypes.number;
    mockArchive.timeStamp.push(new Date(1000));
    mockArchive.timeStamp.push(new Date(2000));
    mockArchive.timeStamp.push(new Date(3000));
    mockArchive.plotValue.push(1);
    mockArchive.plotValue.push(-1);
    mockArchive.plotValue.push(2);
    mockArchive.alarmState.push(0);
    mockArchive.alarmState.push(0);
    mockArchive.alarmState.push(0);

    mount(
      <Plotter
        store={mockStore}
        attribute={mockArchive}
        openPanels={{}}
        theme={mockTheme}
        deriveState={deriveAttributeState}
        doTick
      />
    );
    jest.runTimersToTime(1000 * ARCHIVE_REFRESH_INTERVAL);
    expect(actions.length).toEqual(0);
    jest.runTimersToTime(100 * ARCHIVE_REFRESH_INTERVAL);
    expect(actions.length).toEqual(1);
    expect(actions).toEqual([
      { payload: { path: ['test1', 'attr1'] }, type: MalcolmTickArchive },
    ]);
  });
});

describe('Generic Plotter', () => {
  /*
  let mount;
  let mockArchive;
  let mockStore;
  beforeEach(() => {
    
    mount = createMount();
    mockArchive = {
      timeStamp: new MockCircularBuffer(5),
      alarmState: new MockCircularBuffer(5),
      plotValue: new MockCircularBuffer(5),
      meta: {},
    };
    mockStore = {
      getState: () => {},
      subscribe: () => {},
    };
  });
  */

  it('mouse down in plot area calls startChange function', () => {});

  it('plotly date format function behaves as expected', () => {
    const testDate = new Date(-14258381877);
    expect(plotlyDateFormatter(testDate)).toEqual('1969-07-20 00:20:18.123');
  });

  it('plotly date compare behaves as expected', () => {
    expect(comparePlotlyDateString(new Date(-14258381877), '')).toBeFalsy();
    expect(comparePlotlyDateString('', new Date(-14258381877))).toBeFalsy();

    // identical strings are equal
    expect(
      comparePlotlyDateString(
        '1969-07-20 00:20:18.123',
        '1969-07-20 00:20:18.123'
      )
    ).toBeTruthy();
    // variations after 3rd decimal place are ignored
    expect(
      comparePlotlyDateString(
        '1969-07-20 00:20:18.123',
        '1969-07-20 00:20:18.1234'
      )
    ).toBeTruthy();
    expect(
      comparePlotlyDateString(
        '1969-07-20 00:20:18.1234',
        '1969-07-20 00:20:18.123'
      )
    ).toBeTruthy();
    // trailing zeros do not effect equality
    expect(
      comparePlotlyDateString(
        '1969-07-20 00:20:18.12',
        '1969-07-20 00:20:18.120'
      )
    ).toBeTruthy();
    expect(
      comparePlotlyDateString(
        '1969-07-20 00:20:18.120',
        '1969-07-20 00:20:18.12'
      )
    ).toBeTruthy();
    // differing years returns false
    expect(
      comparePlotlyDateString(
        '1968-07-20 00:20:18.123',
        '1969-07-20 00:20:18.123'
      )
    ).toBeFalsy();
    // differing months returns false
    expect(
      comparePlotlyDateString(
        '1969-08-20 00:20:18.123',
        '1969-07-20 00:20:18.123'
      )
    ).toBeFalsy();
    // differing days returns false
    expect(
      comparePlotlyDateString(
        '1969-07-21 00:20:18.123',
        '1969-07-20 00:20:18.123'
      )
    ).toBeFalsy();
    // differing hours returns false
    expect(
      comparePlotlyDateString(
        '1969-07-20 01:20:18.123',
        '1969-07-20 00:20:18.123'
      )
    ).toBeFalsy();
    // differing minutes returns false
    expect(
      comparePlotlyDateString(
        '1969-07-20 00:21:18.123',
        '1969-07-20 00:20:18.123'
      )
    ).toBeFalsy();
    // differing seconds returns false
    expect(
      comparePlotlyDateString(
        '1969-07-20 00:20:19.123',
        '1969-07-20 00:20:18.123'
      )
    ).toBeFalsy();
    // differing number of milliseconds returns false
    expect(
      comparePlotlyDateString(
        '1969-07-20 00:20:18.121',
        '1969-07-20 00:20:18.120'
      )
    ).toBeFalsy();
  });
});
