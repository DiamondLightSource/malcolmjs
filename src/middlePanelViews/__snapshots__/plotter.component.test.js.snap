// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MethodPlot renders correctly for returns 1`] = `
<PlotlyComponent
  data={
    Array [
      Object {
        "line": Object {
          "dash": undefined,
          "shape": "hv",
        },
        "marker": Object {
          "color": "#f0f0f0",
        },
        "mode": "lines+points",
        "name": "Normal",
        "type": "scatter",
        "visible": true,
        "x": Array [
          1970-01-01T00:00:00.004Z,
          1970-01-01T00:00:00.005Z,
          1970-01-01T00:00:00.006Z,
        ],
        "y": Array [
          undefined,
          undefined,
          null,
        ],
      },
      Object {
        "line": Object {
          "dash": "15px, 5px",
          "shape": "hv",
        },
        "marker": Object {
          "color": "#fff",
        },
        "mode": "lines+points",
        "name": "Warning",
        "type": "scatter",
        "visible": true,
        "x": Array [
          1970-01-01T00:00:00.004Z,
          1970-01-01T00:00:00.005Z,
          1970-01-01T00:00:00.006Z,
        ],
        "y": Array [
          null,
          undefined,
          undefined,
        ],
      },
      Object {
        "line": Object {
          "dash": "10px, 10px",
          "shape": "hv",
        },
        "marker": Object {
          "color": "#ff0000",
        },
        "mode": "lines+points",
        "name": "Major",
        "type": "scatter",
        "visible": true,
        "x": Array [
          1970-01-01T00:00:00.004Z,
          1970-01-01T00:00:00.005Z,
          1970-01-01T00:00:00.006Z,
        ],
        "y": Array [
          null,
          null,
          undefined,
        ],
      },
      Object {
        "line": Object {
          "dash": "5px, 15px",
          "shape": "hv",
        },
        "marker": Object {
          "color": "#0500ff",
        },
        "mode": "lines+points",
        "name": "Disconnected",
        "type": "scatter",
        "visible": false,
        "x": Array [
          1970-01-01T00:00:00.004Z,
          1970-01-01T00:00:00.005Z,
          1970-01-01T00:00:00.006Z,
        ],
        "y": Array [
          null,
          null,
          null,
        ],
      },
    ]
  }
  debug={false}
  divId="plotComponent"
  layout={
    Object {
      "attribute": undefined,
      "autosize": true,
      "datarevision": 0,
      "legend": Object {
        "font": Object {
          "color": "#000",
        },
      },
      "paper_bgcolor": "#424242",
      "parent": undefined,
      "plot_bgcolor": "#424242",
      "xaxis": Object {
        "color": "#000",
        "range": Array [
          1969-12-31T23:59:30.006Z,
          1970-01-01T00:00:00.006Z,
        ],
      },
      "yaxis": Object {
        "color": "#000",
      },
    }
  }
  onRelayout={[Function]}
  style={
    Object {
      "height": "100%",
      "width": "100%",
    }
  }
  useResizeHandler={true}
/>
`;

exports[`MethodPlot renders correctly for takes 1`] = `
<PlotlyComponent
  data={
    Array [
      Object {
        "line": Object {
          "dash": undefined,
          "shape": "hv",
        },
        "marker": Object {
          "color": "#f0f0f0",
        },
        "mode": "lines+points",
        "name": "Normal",
        "type": "scatter",
        "visible": true,
        "x": Array [
          1970-01-01T00:00:00.000Z,
          1970-01-01T00:00:00.002Z,
          1970-01-01T00:00:00.003Z,
        ],
        "y": Array [
          0,
          0,
          null,
        ],
      },
      Object {
        "line": Object {
          "dash": "15px, 5px",
          "shape": "hv",
        },
        "marker": Object {
          "color": "#fff",
        },
        "mode": "lines+points",
        "name": "Warning",
        "type": "scatter",
        "visible": true,
        "x": Array [
          1970-01-01T00:00:00.000Z,
          1970-01-01T00:00:00.002Z,
          1970-01-01T00:00:00.003Z,
        ],
        "y": Array [
          null,
          0,
          0,
        ],
      },
      Object {
        "line": Object {
          "dash": "10px, 10px",
          "shape": "hv",
        },
        "marker": Object {
          "color": "#ff0000",
        },
        "mode": "lines+points",
        "name": "Major",
        "type": "scatter",
        "visible": true,
        "x": Array [
          1970-01-01T00:00:00.000Z,
          1970-01-01T00:00:00.002Z,
          1970-01-01T00:00:00.003Z,
        ],
        "y": Array [
          null,
          null,
          0,
        ],
      },
      Object {
        "line": Object {
          "dash": "5px, 15px",
          "shape": "hv",
        },
        "marker": Object {
          "color": "#0500ff",
        },
        "mode": "lines+points",
        "name": "Disconnected",
        "type": "scatter",
        "visible": false,
        "x": Array [
          1970-01-01T00:00:00.000Z,
          1970-01-01T00:00:00.002Z,
          1970-01-01T00:00:00.003Z,
        ],
        "y": Array [
          null,
          null,
          null,
        ],
      },
    ]
  }
  debug={false}
  divId="plotComponent"
  layout={
    Object {
      "attribute": undefined,
      "autosize": true,
      "datarevision": 0,
      "legend": Object {
        "font": Object {
          "color": "#000",
        },
      },
      "paper_bgcolor": "#424242",
      "parent": undefined,
      "plot_bgcolor": "#424242",
      "xaxis": Object {
        "color": "#000",
        "range": Array [
          1969-12-31T23:59:30.003Z,
          1970-01-01T00:00:00.003Z,
        ],
      },
      "yaxis": Object {
        "color": "#000",
      },
    }
  }
  onRelayout={[Function]}
  style={
    Object {
      "height": "100%",
      "width": "100%",
    }
  }
  useResizeHandler={true}
/>
`;

exports[`attributePlot renders correctly for bool 1`] = `
<PlotlyComponent
  data={
    Array [
      Object {
        "line": Object {
          "dash": undefined,
          "shape": "hv",
        },
        "marker": Object {
          "color": "#f0f0f0",
        },
        "mode": "lines+points",
        "name": "Normal",
        "type": "scatter",
        "visible": true,
        "x": Array [
          1970-01-01T00:00:01.000Z,
          1970-01-01T00:00:02.000Z,
          1970-01-01T00:00:03.000Z,
        ],
        "y": Array [
          1,
          0,
          1,
        ],
      },
      Object {
        "line": Object {
          "dash": "15px, 5px",
          "shape": "hv",
        },
        "marker": Object {
          "color": "#fff",
        },
        "mode": "lines+points",
        "name": "Warning",
        "type": "scatter",
        "visible": false,
        "x": Array [
          1970-01-01T00:00:01.000Z,
          1970-01-01T00:00:02.000Z,
          1970-01-01T00:00:03.000Z,
        ],
        "y": Array [
          null,
          null,
          null,
        ],
      },
      Object {
        "line": Object {
          "dash": "10px, 10px",
          "shape": "hv",
        },
        "marker": Object {
          "color": "#ff0000",
        },
        "mode": "lines+points",
        "name": "Major",
        "type": "scatter",
        "visible": false,
        "x": Array [
          1970-01-01T00:00:01.000Z,
          1970-01-01T00:00:02.000Z,
          1970-01-01T00:00:03.000Z,
        ],
        "y": Array [
          null,
          null,
          null,
        ],
      },
      Object {
        "line": Object {
          "dash": "5px, 15px",
          "shape": "hv",
        },
        "marker": Object {
          "color": "#0500ff",
        },
        "mode": "lines+points",
        "name": "Disconnected",
        "type": "scatter",
        "visible": false,
        "x": Array [
          1970-01-01T00:00:01.000Z,
          1970-01-01T00:00:02.000Z,
          1970-01-01T00:00:03.000Z,
        ],
        "y": Array [
          null,
          null,
          null,
        ],
      },
    ]
  }
  debug={false}
  divId="plotComponent"
  layout={
    Object {
      "attribute": "attr1",
      "autosize": true,
      "datarevision": "test1#attr1#undefined#p:undefined#c:undefined",
      "legend": Object {
        "font": Object {
          "color": "#000",
        },
      },
      "paper_bgcolor": "#424242",
      "parent": "test1",
      "plot_bgcolor": "#424242",
      "xaxis": Object {
        "color": "#000",
        "range": Array [
          1969-12-31T23:59:33.000Z,
          1970-01-01T00:00:03.000Z,
        ],
      },
      "yaxis": Object {
        "color": "#000",
      },
    }
  }
  onRelayout={[Function]}
  style={
    Object {
      "height": "100%",
      "width": "100%",
    }
  }
  useResizeHandler={true}
/>
`;

exports[`attributePlot renders correctly for changing alarm state 1`] = `
<PlotlyComponent
  data={
    Array [
      Object {
        "line": Object {
          "dash": undefined,
          "shape": "hv",
        },
        "marker": Object {
          "color": "#f0f0f0",
        },
        "mode": "lines+points",
        "name": "Normal",
        "type": "scatter",
        "visible": true,
        "x": Array [
          1970-01-01T00:00:01.000Z,
          1970-01-01T00:00:02.000Z,
          1970-01-01T00:00:03.000Z,
        ],
        "y": Array [
          1,
          -1,
          2,
        ],
      },
      Object {
        "line": Object {
          "dash": "15px, 5px",
          "shape": "hv",
        },
        "marker": Object {
          "color": "#fff",
        },
        "mode": "lines+points",
        "name": "Warning",
        "type": "scatter",
        "visible": true,
        "x": Array [
          1970-01-01T00:00:01.000Z,
          1970-01-01T00:00:02.000Z,
          1970-01-01T00:00:03.000Z,
        ],
        "y": Array [
          null,
          -1,
          2,
        ],
      },
      Object {
        "line": Object {
          "dash": "10px, 10px",
          "shape": "hv",
        },
        "marker": Object {
          "color": "#ff0000",
        },
        "mode": "lines+points",
        "name": "Major",
        "type": "scatter",
        "visible": false,
        "x": Array [
          1970-01-01T00:00:01.000Z,
          1970-01-01T00:00:02.000Z,
          1970-01-01T00:00:03.000Z,
        ],
        "y": Array [
          null,
          null,
          null,
        ],
      },
      Object {
        "line": Object {
          "dash": "5px, 15px",
          "shape": "hv",
        },
        "marker": Object {
          "color": "#0500ff",
        },
        "mode": "lines+points",
        "name": "Disconnected",
        "type": "scatter",
        "visible": false,
        "x": Array [
          1970-01-01T00:00:01.000Z,
          1970-01-01T00:00:02.000Z,
          1970-01-01T00:00:03.000Z,
        ],
        "y": Array [
          null,
          null,
          null,
        ],
      },
    ]
  }
  debug={false}
  divId="plotComponent"
  layout={
    Object {
      "attribute": "attr1",
      "autosize": true,
      "datarevision": "test1#attr1#undefined#p:undefined#c:undefined",
      "legend": Object {
        "font": Object {
          "color": "#000",
        },
      },
      "paper_bgcolor": "#424242",
      "parent": "test1",
      "plot_bgcolor": "#424242",
      "xaxis": Object {
        "color": "#000",
        "range": Array [
          1969-12-31T23:59:33.000Z,
          1970-01-01T00:00:03.000Z,
        ],
      },
      "yaxis": Object {
        "color": "#000",
      },
    }
  }
  onRelayout={[Function]}
  style={
    Object {
      "height": "100%",
      "width": "100%",
    }
  }
  useResizeHandler={true}
/>
`;

exports[`attributePlot renders correctly for number 1`] = `
<PlotlyComponent
  data={
    Array [
      Object {
        "line": Object {
          "dash": undefined,
          "shape": "hv",
        },
        "marker": Object {
          "color": "#f0f0f0",
        },
        "mode": "lines+points",
        "name": "Normal",
        "type": "scatter",
        "visible": true,
        "x": Array [
          1970-01-01T00:00:01.000Z,
          1970-01-01T00:00:02.000Z,
          1970-01-01T00:00:03.000Z,
        ],
        "y": Array [
          1,
          -1,
          2,
        ],
      },
      Object {
        "line": Object {
          "dash": "15px, 5px",
          "shape": "hv",
        },
        "marker": Object {
          "color": "#fff",
        },
        "mode": "lines+points",
        "name": "Warning",
        "type": "scatter",
        "visible": false,
        "x": Array [
          1970-01-01T00:00:01.000Z,
          1970-01-01T00:00:02.000Z,
          1970-01-01T00:00:03.000Z,
        ],
        "y": Array [
          null,
          null,
          null,
        ],
      },
      Object {
        "line": Object {
          "dash": "10px, 10px",
          "shape": "hv",
        },
        "marker": Object {
          "color": "#ff0000",
        },
        "mode": "lines+points",
        "name": "Major",
        "type": "scatter",
        "visible": false,
        "x": Array [
          1970-01-01T00:00:01.000Z,
          1970-01-01T00:00:02.000Z,
          1970-01-01T00:00:03.000Z,
        ],
        "y": Array [
          null,
          null,
          null,
        ],
      },
      Object {
        "line": Object {
          "dash": "5px, 15px",
          "shape": "hv",
        },
        "marker": Object {
          "color": "#0500ff",
        },
        "mode": "lines+points",
        "name": "Disconnected",
        "type": "scatter",
        "visible": false,
        "x": Array [
          1970-01-01T00:00:01.000Z,
          1970-01-01T00:00:02.000Z,
          1970-01-01T00:00:03.000Z,
        ],
        "y": Array [
          null,
          null,
          null,
        ],
      },
    ]
  }
  debug={false}
  divId="plotComponent"
  layout={
    Object {
      "attribute": "attr1",
      "autosize": true,
      "datarevision": "test1#attr1#undefined#p:undefined#c:undefined",
      "legend": Object {
        "font": Object {
          "color": "#000",
        },
      },
      "paper_bgcolor": "#424242",
      "parent": "test1",
      "plot_bgcolor": "#424242",
      "xaxis": Object {
        "color": "#000",
        "range": Array [
          1969-12-31T23:59:33.000Z,
          1970-01-01T00:00:03.000Z,
        ],
      },
      "yaxis": Object {
        "color": "#000",
      },
    }
  }
  onRelayout={[Function]}
  style={
    Object {
      "height": "100%",
      "width": "100%",
    }
  }
  useResizeHandler={true}
/>
`;
